File: 0064-minimum-path-sum/0064-minimum-path-sum.java
class Solution {
   public int minPathSum(int[][] grid) {
       return minPathSumHelper(grid,0,0);
    }

    private int minPathSumHelper(int[][] grid,int row, int col){
        if (row>=grid.length || col>=grid[0].length) return Integer.MAX_VALUE;
        else if (row==grid.length-1 && col==grid[0].length-1) return grid[row][col];
        else if (grid[row][col] < 101) grid[row][col] += 101 + Math.min(minPathSumHelper(grid, row+1, col), minPathSumHelper(grid, row, col+1));

        return grid[row][col] - 101;
    }
}

File: 0064-minimum-path-sum/0064-minimum-path-sum.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])

        @cache
        def bt(i, j):
            if i == m - 1 and j == n - 1:
                return grid[i][j]
            if j == n - 1 and 0 <= i < m - 1:
                return grid[i][j] + bt(i + 1, j)
            if i == m - 1 and 0 <= j < n - 1:
                return grid[i][j] + bt(i, j + 1)
            if 0 <= j < n - 1 and 0 <= i < m - 1:
                return grid[i][j] + min(bt(i + 1, j), bt(i, j + 1))
        return bt(0, 0)
        